vyper/old_codegen/expr.py

```diff
@@ -317,6 +317,36 @@ def parse_Name(self):
             return LLLnode.from_list(
                 [obj], typ=BaseType(typ, is_literal=True), pos=getpos(self.expr)
             )
+        elif self.expr._metadata["type"].is_immutable:
+            # immutable variable
+            # need to handle constructor and outside constructor
+            var = self.context.globals[self.expr.id]
+            is_constructor = self.expr.get_ancestor(vy_ast.FunctionDef).get("name") == "__init__"
+            if is_constructor:
+                memory_loc = self.context.new_variable(f"#immutable_{self.expr.id}", var.typ)
+                data_offset = self.expr._metadata["type"].position.offset
+                var._metadata = {"memory_loc": memory_loc, "data_offset": data_offset}
+                return LLLnode.from_list(
+                    memory_loc,
+                    typ=var.typ,
+                    location="memory",
+                    pos=getpos(self.expr),
+                    annotation=self.expr.id,
+                    mutable=True,
+                )
+            else:
+                immutable_section_size = sum(
+                    [imm.size * 32 for imm in self.context.globals.values() if imm.is_immutable]
+                )
+                offset = self.expr._metadata["type"].position.offset
+                return LLLnode.from_list(
+                    ["sub", "codesize", immutable_section_size - offset],
```

Comment:
This is interesting. I wonder if we can get the location of the data section into the runtime code during deploy time.

Reply:
Here's an idea: our assembly-to-evm compilation phase has a step which resolves symbols to locations. Since the locations of immutables are statically known (no matter what value they are set to), we can insert symbols into the LLL which directly precede blank space. Basically, _sym_foo doesn't actually take any space in the code. So the runtime data section itself would have something like

```
_sym_datasection_start
BLANK
_sym_immutable_uint256
JUMPDEST... <32 JUMPDESTs>
_sym_immutable_bytes6
JUMPDEST... <32+6 jumpdests>
```
Then your code for copying items to LLL would stay the same (it would just overwrite all the jumpdests), but we would know the locations statically.

Reply:
OK let's add a TODO here for now: `TODO: resolve code offsets for immutables at compile time`